include/kodlab_mjbots_sdk/lcm_subscriber.h

```diff
@@ -1,85 +1,149 @@
-  /*!
-   * @brief callback function when msg is received. Copies the message to m_data
-   * @param rbuf unknown, but not used
-   * @param chan channel name
-   * @param msg ptr to the incoming message data
+  /**
+   * @brief Data for the most-recently received message.
+   */
+  Message data;
+
+  /**
+   * @brief Mutex for the message handler.
+   */
+  std::mutex mutex_;
+
+  /**
+   * @brief Constructs an `LcmMessageHandler` and ensures the mutex is unlocked.
    */
-  void HandleMsg(const lcm::ReceiveBuffer *rbuf,
-                 const std::string &chan,
-                 const MessageClass *msg);
+  LcmMessageHandler()
+  {
+    mutex_.unlock();  // ensures mutex is unlocked
+  }
+
+  /**
+   * @brief Callback function for when an LCM message is received.
+   * @details This function serves as a callback for incoming LCM messages on
+   * `channel`.  It copies the decoded message data to `data`, and sets
+   * `new_message` to `true`.
+   * @param rbuf LCM recieve buffer containing raw bytes and timestamp
+   * @param channel channel name
+   * @param msg pointer to the incoming message data
+   */
+  void HandleMessage(const lcm::ReceiveBuffer *rbuf,
+                     const std::string &channel,
+                     const Message *msg)
+  {
+    mutex_.lock();
+    data = *msg;
+    new_message = true;
+    mutex_.unlock();
+  }
+};
+
+/**
+ * @brief LCM subscriber capable of subscribing to multiple channels
+ * @todo Implement subscription removal (i.e., unsubscribing)
+ */
+class LcmSubscriber : public AbstractRealtimeObject
+{
+public:
 
   /*!
-   * @brief subscribes to the lcm channel using handle message and handles ctrl c detection
+   * @brief Constructor an lcm subscriber
+   * @param realtime_priority realtime priority in range [1, 99]
+   * @param cpu cpu for this process
    */
-  void Run() override;
+  LcmSubscriber(int realtime_priority, int cpu)
+  {
+    cpu_ = cpu;
+    realtime_priority_ = realtime_priority;
+    Start();
+  }
 
-  std::string channel_name_;
+  /**
+   * @brief Add a new LCM channel to the the LCM object's subscriptions
+   * @tparam Message LCM message type for channel
+   * @param channel_name channel name
+   * @param handler `LcmMessageHandler`-inherited message handler object
+   * @return generated `lcm::Subscription` object
+   */
+  template<class Message>
+  lcm::Subscription *AddSubscription(std::string channel_name,
+                                     LcmMessageHandler<Message> &handler)
+  {
+    auto sub = lcm_.subscribe(channel_name,
+                              &LcmMessageHandler<Message>::HandleMessage,
+                              &handler);
+    subs_.emplace(channel_name, sub);
+    return sub;
+  }
+
+  /**
+   * @brief Accessor for `lcm::Subscription` objects
+   * @param channel_name channel name
+   * @return `lcm::Subscription` object on corresponding channel
+   */
+  [[nodiscard]] const lcm::Subscription *get_subscription(const std::string &channel_name) const
```

Comment:
nodiscard is growing on me, even though its ugly haha

Reply:
Same here, ugly but it does some good!